odoo/orm/models.py

```diff
@@ -5157,8 +5157,40 @@ def exists(self) -> Self:
             return self
         query = Query(self.env, self._table, self._table_sql)
         query.add_where(SQL("%s IN %s", SQL.identifier(self._table, 'id'), tuple(ids)))
-        self.env.cr.execute(query.select())
-        valid_ids = set([r[0] for r in self._cr.fetchall()] + new_ids)
+        real_ids = (id_ for [id_] in self.env.execute_query(query.select()))
+        valid_ids = {*real_ids, *new_ids}
+        return self.browse(i for i in self._ids if i in valid_ids)
+
+    @api.private
+    def lock_for_update(self, *, allow_referencing: bool = False) -> Self:
+        """ Grab an exclusive write-lock to the rows with the given ids.
+
+        This avoids blocking processing on the records due to concurrent
+        modifications. If all records couldn't be locked, a `LockFailedError`
+        exception is risen.
+
+        :param allow_referencing: Acquire a row lock which allows for other
+            transactions to reference this record. Use only when modifying
+            values that are not identifiers.
+        :return: Locked records with prefetch
+        :raises: ``LockFailedError`` when some records could not be locked
+        """
+        new_ids, ids = partition(lambda i: isinstance(i, NewId), self._ids)
+        if not ids:
+            return self
+        query = Query(self.env, self._table, self._table_sql)
+        query.add_where(SQL("%s IN %s", SQL.identifier(self._table, 'id'), tuple(ids)))
+        if allow_referencing:
+            lock_sql = SQL("FOR NO KEY UPDATE")
+        else:
+            lock_sql = SQL("FOR UPDATE")
+        # Use SKIP LOCKED instead of NOWAIT because the later aborts the
+        # transaction and we do not want to use SAVEPOINTS.
+        valid_ids = self.env.execute_query(SQL("%s %s SKIP LOCKED", query.select(), lock_sql))
+        valid_ids = {id_ for [id_] in valid_ids}
```

Comment:
```suggestion
        # Use SKIP LOCKED instead of NOWAIT because the later aborts the
        # transaction and we do not want to use SAVEPOINTS.
        if allow_referencing:
            lock_sql = SQL("FOR NO KEY UPDATE SKIP LOCKED")
        else:
            lock_sql = SQL("FOR UPDATE SKIP LOCKED")
        rows = self.env.execute_query(SQL("%s %s", query.select(), lock_sql))
        valid_ids = {id_ for [id_] in rows}
```